addon/components/table-of-contents-plugin/ember-nodes/table-of-contents.ts

```diff
@@ -19,11 +19,38 @@ export default class TableOfContentsComponent extends Component<EmberNodeArgs> {
     return this.args.controller;
   }
 
+  get tableOfContentsRange() {
+    let result: { from: number; to: number } | undefined;
+    this.controller.mainEditorState.doc.descendants((node, pos) => {
+      if (node.type === this.controller.schema.nodes['table_of_contents']) {
+        result = { from: pos, to: pos + node.nodeSize };
+      }
+      return !result;
+    });
+    return result;
+  }
+
+  @action
+  onUpdate() {
+    if (this.tableOfContentsRange) {
+      const { from, to } = this.tableOfContentsRange;
+      this.controller.withTransaction((tr) => {
+        const node = this.args.node;
+        const attrs = node.attrs;
+        const newAttrs = { ...attrs, ...this.outline };
+        const newNode = node.type.create(newAttrs, node.content, node.marks);
+        tr.replaceWith(from, to, newNode);
```

Comment:
any reason for not using `setNodeMarkup` here?
https://prosemirror.net/docs/ref/#transform.Transform.setNodeMarkup
looks to me you don't even need to calculate the range then and can just use the `args.getPosition` function